File: 400/437 - Path Sum III.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    
    public int pathSum(TreeNode root, int sum) {
        return helper(root, sum, 0, new ArrayList<Integer>());
    }
    
    private int helper(TreeNode root, int sum, int curr, List<Integer> list) {
        int ret = 0, i, temp;
        
        if (root == null)
            return 0;   
        list.add(root.val);
        
        curr += root.val;
        if (curr == sum)
            ++ret;
        
        temp = curr;
        for (i = 0; i < list.size() - 1; ++i) {
            temp -= list.get(i);
            if (temp == sum)
                ++ret;
        }
        
        ret += helper(root.left, sum, curr, list) + helper(root.right, sum, curr, list);
        list.remove(list.size() - 1);
        
        return ret;
    }
}

File: csharp/437.path-sum-iii.cs
/*
 * @lc app=leetcode id=437 lang=csharp
 *
 * [437] Path Sum III
 */
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int x) { val = x; }
 * }
 */

using System.Collections.Generic;

public class Solution {

    private static int PathSumRecurse(TreeNode curr, IDictionary<int, int> count, int target, int sum) {
        if (curr == null) {
            return 0;
        }
        sum += curr.val;
        int ret = count.ContainsKey(sum - target) ? count[sum - target] : 0;
        count[sum] = (count.ContainsKey(sum) ? count[sum] : 0) + 1;
        ret += PathSumRecurse(curr.left, count, target, sum) + PathSumRecurse(curr.right, count, target, sum);
        --count[sum];
        return ret;
    }

    public int PathSum(TreeNode root, int sum) {
        var count = new Dictionary<int, int>();
        count[0] = 1;
        return PathSumRecurse(root, count, sum, 0);
    }
}

